src/wallet/rpcwallet.cpp

```diff
@@ -387,23 +387,22 @@ UniValue listaddresses(const UniValue& params, bool fHelp)
     // This will include any address that has been generated by this wallet.
     for (const std::pair<CTxDestination, CAddressBookData>& item : pwalletMain->mapAddressBook) {
         t_generated_dests.insert(item.first);
+        auto script = GetScriptForDestination(item.first);
+        if (pwalletMain->HaveWatchOnly(script)) {
+            t_watchonly_dests.insert(item.first);
+        }
```

Comment:
```suggestion
        auto script = GetScriptForDestination(item.first);
        if (pwalletMain->HaveWatchOnly(script)) {
            t_watchonly_dests.insert(item.first);
        } else {
            t_generated_dests.insert(item.first);
        }
```

Reply:
We discussed this in a pairing, and decided that this should instead switch on `GetSourceForPaymentAddress(item.first)`. We know the only two cases we need to handle are `Random` (which whill include `Imported` because we can't distinguish those) and `ImportedWatchOnly`; no other transparent address kind gets added to the addressbook.